### PR TITLE
Improve performance of the impute function.

### PR DESCRIPTION
**Pull request by F-A**
Improve performance of the functions:
- utilities.dataframe_function.get_range_values_per_column(df)
- utilities.dataframe_function.impute(df_impute)
More specifically: apply the impute function directly on numpy array to improve computation time.

Now the impute function runs in 109ms (60 samples, 14256 features (or columns) ).

Note: I did not impoved the performance of impute_dataframe_range(...) since it would have been to much of a hassle to implement all the checks in that function, e.g. in case the min/max/median values of each columns are not present. In our case we call the get_range_values_per_column just before so these checks are not necessary.
So I just reimplemented the function impute_dataframe_rage directly in the impute function. This is less modular. Maybe you could pack this code in a new impute_dataframe_range function.

Solve the issue #123 .

### CODE REVIEW

File: tsfresh/utilities/dataframe_functions.py [hunk 1]
```diff
@@ -55,7 +55,30 @@ def impute(df_impute):
...
+
+    # NaN -> median
+    columns = df_impute.columns
+    indices = np.nonzero(np.isnan(df_impute.values))
+    if len(indices[0]) > 0:
+        newValues = [ col_to_median[ columns[i] ] for i in indices[1]]
+        df_impute.values[ indices ] = newValues
+
+    # Ensure a type of "np.float64"
+    df_impute.astype(np.float64)
```
Comment (MaxBenChrist) [inline tsfresh/utilities/dataframe_functions.py]: you have to set `copy=False` here

Comment (MaxBenChrist) [inline tsfresh/utilities/dataframe_functions.py]: otherwise it will just return a copy

**Comment by MaxBenChrist** [comment]
>Note: I did not impoved the performance of impute_dataframe_range(...) since it would have been to much of a hassle to implement all the checks in that function, e.g. in case the min/max/median values of each columns are not present.

I think the only check that is missing in your version is if there are no finite values at all

``` python

            if len(finite_values_in_column) == 0:
                _logger.warning(
                    "The replacement column {} did not have any finite values. Filling with zeros.".format(column_name))
                df_impute[column_name] = [0] * len(column)
                continue
```

**Comment by F-A** [comment]
The check in case there no finite values in done in the "get_range_values_per_column" function. Line 173-175. But it does not log anything.

**Comment by MaxBenChrist** [comment]
Ah yes you are right, couldn't you do something like

``` python
if len(masked.mask.sum(axis=0) == masked.data.shape[0]) > 0
    logging message")
```

**Comment by F-A** [comment]
The log message was previously done in the impute function directly. In this function I do not create the masked array.
I could add the warning in the "get_range_values_per_column" where I have the masked array but this does make less sense...

**Comment by jneuff** [comment]
@F-A Nice work!

The impute function is provided as a default for users who don't know which impute function to choose (we currently have two: impute_dataframe_range and impute_zero). From the semantics you're proposal is an improvement of the impute_dataframe_range function. So that's the function it should replace and impute should remain just a wrapper.

**Comment by jneuff** [comment]
Ah, and while we're at it: I'd prefer the impute function to return the altered dataframe. Currently we're inconsistent about this.

**Comment by MaxBenChrist** [comment]
> Ah, and while we're at it: I'd prefer the impute function to return the altered dataframe. Currently we're inconsistent about this.

The problem is that you create (temporary) copys of the dataframe if you impute that way. For big dataframes this could result in out of memory problems.

**Comment by F-A** [comment]
Rewriting the  impute_dataframe_range with its current behavior would take me too much time: it assumes the max/min and median dict given in parameters can be incomplete, which results in a lot of column-wise checks. The impute function directly compute these dictionaries, so we know they are complete, which make the implementation easier.

As I mentioned earlier, I do not have much time to accord to tsfresh. I just wanted to test it on my data and I fixed this issue to do so. This PR correspond to my fix.
You can accept it if you want and improve the functions "impute_dataframe_range" and "impute_zero" in the same spirit later on. But this will be too much for me right now, sorry.

**Comment by MaxBenChrist** [comment]
Thank you. no worries, we will take over from here. 

May I push some changes to your master branch to update this pr?

**Comment by F-A** [comment]
Yes of course. I'll grant you access to it when I am in front of a computer.

On 6 Jan 2017 23:46, "Maximilian Christ" <notifications@github.com> wrote:

Thank you. no worries, we will take over from here.

May I push some changes to your master branch to update this pr?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/blue-yonder/tsfresh/pull/135#issuecomment-271031182>,
or mute the thread

**Comment by MaxBenChrist** [comment]
> Yes of course. I'll grant you access to it when I am in front of a computer.

Perfect, let me know when you arranged it.

**Comment by jneuff** [comment]
> The problem is that you create (temporary) copys of the dataframe if you impute that way. For big dataframes this could result in out of memory errors. I think we should do all imputing inplace

We could do it in place and return the dataframe. That way, we won't have to change the function if we later decide to make our data flow more explicit as some kind of pipeline.

@F-A No, worries. Thanks for contributing!

I just noticed, that we currently have quite some code duplication between `impute_dataframe_range` and `get_range_values_per_column`. We should simplify that anyway. This PR seems like a good opportunity.

**Comment by F-A** [comment]
Ok I added both of you as collabortors for my fork. You should be able to push then.

**Comment by MaxBenChrist** [comment]
Thanks, let's use this opportunity to refactor the module. I will have a look at it during the next days.

**Comment by nils-braun** [comment]
@MaxBenChrist Is there news on this branch and PR? Or do you want to tell us your plans for refactoring - so we can help out? ;-)

**Comment by MaxBenChrist** [comment]
I do not have a fixed plan yet but some aspects I want to tackle

- remove code duplication between impute_dataframe_range and get_range_values_per_column
- maybe remove impute_dataframe_zero
- check if we still need get_range_values_per_column if we use the masks
- ...

Maybe I will find some time during the next days to tackle this.

**Comment by MaxBenChrist** [comment]
I finally found some hours to spend on this. I finished refactoring the dataframe_functions module.

However, I found that the usage of the impute functions in the transformers is not right. I will try to fix that as well

**Comment by MaxBenChrist** [comment]
Well, I finished everything. If either @nils-braun or @jneuff gives his go, we can merge this thing.

**Comment by jneuff** [comment]
Thanks @F-A and @MaxBenChrist !

**Comment by MaxBenChrist** [comment]
Awesome! We are on the way to make tsfresh blazing fast ;)

**Comment by F-A** [comment]
Nice refactoring! I'm glad my original investigations helped you :)

**Pull request merged**